Refactored perception pipeline tutorial from ros1 to ros 2 (#906)
moveit/moveit2_tutorials: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst

Perception Pipeline Tutorial
==================================

Introduction
------------
MoveIt serves sensor plugins for utilizing 3D sensor data in planning pipeline. When it's ran the demo of this tutorial, it should be seen the following rviz screen.

.. raw:: html

>         <iframe width="560" height="315" src="https://www.youtube.com/embed/bpMHFxcbRXg" title="MoveIt Perception Pipeline Demo" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

Reviewer: Very nice that you recorded a video! Instead of linking a youtube video here, wouldn't a gif work too?
Additional, I think it would be nice if the video just contains the relevant video and not your entire screen. I use `kazam` to record videos and it is very straight forward to capture a window only

------------
MoveIt serves sensor plugins for utilizing 3D sensor data in planning pipeline. When it's ran the demo of this tutorial, it should be seen the following rviz screen.

MoveIt allows for seamless integration of 3D sensors using `Octomap <http://octomap.github.io/>`_.
Once properly configured, you should see something like this in rviz:
.. raw:: html

.. image:: perception_configuration_demo.png
   :width: 700px
        <iframe width="560" height="315" src="https://www.youtube.com/embed/bpMHFxcbRXg" title="MoveIt Perception Pipeline Demo" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

Getting Started
---------------
If you haven't already done so, make sure you've completed the steps in :doc:`Getting Started </doc/tutorials/getting_started/getting_started>`.

> Configuration

Reviewer: Why are you deleting all this, is it outdated?

Author: I actually didn't delete it. I just made some modifications. You can find the relavant redirection in this link .

https://github.com/CihatAltiparmak/moveit2_tutorials/blob/e7eec45b14d360b8e8ffe2471b38f22e9f7e05cf/doc/examples/perception_pipeline/perception_pipeline_tutorial.rst?plain=1#L15-L18

> In this tutorial, you can use `previously recorded 3D pointcloud data <https://drive.google.com/file/d/1fPtDAtJKIiw2gpFOOwA2TrPZOfFU053W/view?usp=sharing>`_ or you can record your own bag file. For recording bag, firstly, it can be both ran ``depth_camera_envrionment.launch.py`` and recorded bag using following commands.

Reviewer: I wouldn't use your personal cloud. We could add the file to an additional repository with git lfs to store it. How big is it?

Author: Agree, but unfortunatelly the bag size is approximately 56MB. I will try to record this bag file in lower size. Do you have any suggestion about it? I'm not sure git lfs works here.

Author: Now it's not necessary. We are about to move bag file into moveit_benchmark_resources

:moveit1:

..
   Once updated for MoveIt 2, remove all lines above title (including this comment and :moveit1: tag)

Perception Pipeline Tutorial
============================
==================================

Introduction
------------
MoveIt serves sensor plugins for utilizing 3D sensor data in planning pipeline. When it's ran the demo of this tutorial, it should be seen the following rviz screen.

MoveIt allows for seamless integration of 3D sensors using `Octomap <http://octomap.github.io/>`_.
Once properly configured, you should see something like this in rviz:
.. raw:: html

.. image:: perception_configuration_demo.png
   :width: 700px
        <video width="700px" controls="true" autoplay="true" loop="true">
            <source src="../../../_static/videos/perception_pipeline_demo.webm" type="video/webm">
            MoveIt Perception Pipeline Demo
        </video>

Getting Started
---------------
If you haven't already done so, make sure you've completed the steps in :doc:`Getting Started </doc/tutorials/getting_started/getting_started>`.

Configuration
-------------

> In this section, we will walk through configuring the 3D sensors on your robot with MoveIt. The primary component in MoveIt that deals with 3D perception is the Occupancy Map Updater. The updater uses a plugin architecture to process different types of input. The currently available plugins in MoveIt are:

Reviewer: This section shouldn't be removed

Author: Added to this part . https://github.com/CihatAltiparmak/moveit2_tutorials/blob/bc1504d3d6c13e17c19bf506fe70358bbc83a100/doc/examples/perception_pipeline/perception_pipeline_tutorial.rst?plain=1#L21-L24

* The Depth Image Occupancy Map Updater: which can take as input Depth Images (``sensor_msgs/Image``)

To use the Occupancy Map Updater, it is only necessary to set the appropriate parameters on the ROS parameter server and to call ``startWorldGeometryMonitor`` from your ``PlanningSceneMonitor``.  This latter step is performed automatically when using Move Group functionality like in this tutorial's example, so in that case it is only necessary to set the parameters for the octomap and octomap updater.

YAML Configuration file (Point Cloud)
+++++++++++++++++++++++++++++++++++++

We will have to generate a YAML configuration file for configuring the 3D sensors. Please see :panda_codedir:`this example file<config/sensors_kinect_pointcloud.yaml>` for processing point clouds.

Save this file in the config folder in the robot's moveit_config package with name "sensors_kinect_pointcloud.yaml": ::

 sensors:
   - sensor_plugin: occupancy_map_monitor/PointCloudOctomapUpdater
     point_cloud_topic: /camera/depth_registered/points
     max_range: 5.0
     point_subsample: 1
     padding_offset: 0.1
     padding_scale: 1.0
     max_update_rate: 1.0
     filtered_cloud_topic: filtered_cloud

> **The general parameters are:**

Reviewer: Can we keep the param descriptions too? You could merge them as a comment with the `sensors_3d.yaml: ::` description

Author: I have added to https://github.com/CihatAltiparmak/moveit2_tutorials/blob/bc1504d3d6c13e17c19bf506fe70358bbc83a100/doc/examples/perception_pipeline/perception_pipeline_tutorial.rst?plain=1#L90-L108 . I will also add for depth map plugin.

* *sensor_plugin*: The name of the plugin that we are using.
* *max_update_rate*: The octomap representation will be updated at rate less than or equal to this value.

**Parameters specific to the Point cloud updater are:**

* *point_cloud_topic*: This specifies the topic to listen on for a point cloud.

* *max_range*: (in m) Points further than this will not be used.

* *point_subsample*: Choose one of every *point_subsample* points.

* *padding_offset*: The size of the padding (in cm).

* *padding_scale*: The scale of the padding.

* *filtered_cloud_topic*: The topic on which the filtered cloud will be published (mainly for debugging). The filtering cloud is the resultant cloud after self-filtering has been performed.

> YAML Configuration file (Depth Map)

Reviewer: We can drop the depth map since it is not covered in this tutorial anyways I guess

Author: Added

> Even though you haven't completed :doc:`Getting Started </doc/tutorials/getting_started/getting_started>` ever, you can still run this tutorial's demo. Whereas, it is suggested to start with steps in :doc:`Getting Started </doc/tutorials/getting_started/getting_started>` for understanding what's going on in this tutorial.

Reviewer: ```suggestion
Even though you haven't completed :doc:`Getting Started </doc/tutorials/getting_started/getting_started>` ever, you can still run this tutorial's demo. But, it is suggested to start with steps in :doc:`Getting Started </doc/tutorials/getting_started/getting_started>` for better understanding what's going on in this tutorial.
```

> Get The Necessary 3D Pointcloud Data For Octomap Creation

Reviewer: ```suggestion
How to create 3D Pointcloud Data for Octomap Creation
```
Just a suggestion

> In this tutorial, you can use `previously recorded 3D pointcloud data <https://drive.google.com/file/d/1fPtDAtJKIiw2gpFOOwA2TrPZOfFU053W/view?usp=sharing>`_ or you can record your own bag file. For recording bag, firstly, it can be both run ``depth_camera_envrionment.launch.py`` and recorded bag using following commands.

Reviewer: There's a typo in `depth_camera_envrionment.launch.py` and I don't understand the second sentence. Do you mind rewording it?

Author: You're absolutely right. I reworded like this 

```
For recording bag, firstly, we can run ``depth_camera_envrionment.launch.py`` file and then record the bag using following commands.
```

Octomap Configuration
^^^^^^^^^^^^^^^^^^^^^
You will also need to configure the `Octomap <http://octomap.github.io/>`_ by adding the following lines into the *sensor_manager.launch*: ::
Even though you haven't completed :doc:`Getting Started </doc/tutorials/getting_started/getting_started>` ever, you can still run this tutorial's demo. Whereas, it is suggested to start with steps in :doc:`Getting Started </doc/tutorials/getting_started/getting_started>` for understanding what's going on in this tutorial.

 <param name="octomap_frame" type="string" value="odom_combined" />
 <param name="octomap_resolution" type="double" value="0.05" />
 <param name="max_range" type="double" value="5.0" />
Get The Necessary 3D Pointcloud Data For Octomap Creation
---------------------------------------------------------
In this tutorial, you can use `previously recorded 3D pointcloud data <https://drive.google.com/file/d/1fPtDAtJKIiw2gpFOOwA2TrPZOfFU053W/view?usp=sharing>`_ or you can record your own bag file. For recording bag, firstly, it can be both run ``depth_camera_envrionment.launch.py`` and recorded bag using following commands.

> MoveIt uses an octree-based framework to represent the world around it. The *Octomap* parameters above are configuration parameters for this representation:

Reviewer: Can we keep this sentence? Maybe move it to the introduction

Author: Added in https://github.com/CihatAltiparmak/moveit2_tutorials/blob/bc1504d3d6c13e17c19bf506fe70358bbc83a100/doc/examples/perception_pipeline/perception_pipeline_tutorial.rst?plain=1#L57-L62

> In next step, it will be used this recorded bag file for octomap creation.

Reviewer: ```suggestion
In next step, we will use the recorded bag file to create an octomap.
```
Rewording suggestion

> Now whatever you get bag file, you will see ``/camera_1/points``, ``/camera_2/points``, ``/tf`` and ``/tf_static`` in both way when played bag file. It should be created following config file for MoveIt to process these pointcloud topics in planning pipeline. You can also go to :codedir:`here <examples/perception_pipeline/config/sensors_3d.yaml>` for seeing all ``sensors_3d.yaml`` config file on Github.

Reviewer: Do you mind rewording this to be a bit more precise too? I am not sure if I understand it correctly.

Author: I reworded it here and i added another part based on your suggestion.

https://github.com/CihatAltiparmak/moveit2_tutorials/blob/bc1504d3d6c13e17c19bf506fe70358bbc83a100/doc/examples/perception_pipeline/perception_pipeline_tutorial.rst?plain=1#L57-L62

>         sensor_plugin: occupancy_map_monitor/PointCloudOctomapUpdater

Reviewer: How much work would it be to add a depthcamera example here instead of the second Pointcloud updater (and can you briefly explain why we need two if that's required)?

Author: It's not necessary to use two pointcloud, These pointclouds show both two sides of table. We can turn second plugin into depth camera plugin. When i completed, i will tag you.

Author: I've added. You can find some configurations here. 

https://github.com/CihatAltiparmak/moveit2_tutorials/blob/c5383b570f8e03157165f878dc67f27c6ec3df28/doc/examples/perception_pipeline/config/sensors_3d.yaml#L1-L22

Perception Pipeline Tutorial

> ==================================

Reviewer: When building locally, I get a warning when this does not match the number of heading letters 
```suggestion
============================
```

> 1 Introduction

Reviewer: Remove the numbers before the headings to match the style of the other tutorials
```suggestion
Introduction
```
